### PR TITLE
🌱   Upgrade to use latest Kind version

### LINKED ISSUE #6264: Upgrade to kind v0.12.0
Hey folks,

kind v0.12.0 was just recently released and alongside new node images based on the new kind version.

Let's migrate to the new version.

We have to do **at least**:
* [ ] upgrade kind (hack/ensure-kind.sh, test/go.mod)
* [ ] upgrade our e2e test config on main and for main (in test-infra) to use new kind images (https://github.com/kubernetes-sigs/kind/releases/tag/v0.12.0)

/kind feature
/area provider/docker

### PR DESCRIPTION
**Pull request by DiptoChakrabarty**
**What this PR does / why we need it**: This updates to make use of latest kind v0.12.0 version 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6264

### CODE REVIEW

**Comment by sbueringer** [comment]
Looks like you have to run `make generate-modules`

There should be a few more Kubernetes (and thus kind) versions in docker.yaml which we should bump to the latest kind-published image versions

File: test/e2e/config/docker.yaml [hunk 1]
```diff
@@ -199,10 +199,10 @@ variables:
   # allowing the same e2e config file to be re-used in different Prow jobs e.g. each one with a K8s version permutation.
   # The following Kubernetes versions should be the latest versions with already published kindest/node images.
   # This avoids building node images in the default case which improves the test duration significantly.
-  KUBERNETES_VERSION_MANAGEMENT: "v1.23.3"
-  KUBERNETES_VERSION: "v1.23.3"
-  KUBERNETES_VERSION_UPGRADE_FROM: "v1.22.4"
-  KUBERNETES_VERSION_UPGRADE_TO: "v1.23.3"
+  KUBERNETES_VERSION_MANAGEMENT: "v1.23.4"
+  KUBERNETES_VERSION: "v1.23.4"
+  KUBERNETES_VERSION_UPGRADE_FROM: "v1.23.3"
```
Comment (sbueringer) [inline test/e2e/config/docker.yaml:204]: This should be 1.22.7

Comment (DiptoChakrabarty) [reply test/e2e/config/docker.yaml:204]: @sbueringer `KUBERNETES_VERSION_UPGRADE_FROM=1.22.7` is this correct

**Comment by sbueringer** [comment]
Sorry missed the last push force notification somehow

/lgtm
/test pull-cluster-api-e2e-full-main

/hold 
for full e2e tests

/assign @fabriziopandini

**Pull request merged**